src/libjson/json.c

```diff
@@ -0,0 +1,1736 @@
+    }
+
+    while (*src)
+    {
+        src++;
+        n++;
+    }
+
+    return n;
+}
+
+//
+// If c is a digit character:
+//     then: _digit[c] yields the integer value for that digit character.
+//     else: _digit[c] yields 0xFF.
+//
+// Digit characters fall within these ranges: ['0'-'9'] and ['A'-'Z'].
+//
+// Examples:
+//     _digit['9'] => 9
+//     _digit['A'] => 10
+//     _digit['Z'] => 35
+//     _digit['?'] => 0xFF
+//
+static const unsigned char _digit[256] = {
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10,
+    0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c,
+    0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14,
+    0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20,
+    0x21, 0x22, 0x23, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xFF, 0xFF, 0xFF, 0xFF,
+};
+
+/* Return true if c is a digit character with the given base */
+static int _isdigit2(char c, int base)
+{
+    return _digit[(unsigned char)c] < base;
+}
+
+static long int _strtol(const char* nptr, char** endptr, int base)
+{
+    const char* p;
+    unsigned long x = 0;
+    int negative = 0;
+
+    if (endptr)
+        *endptr = (char*)nptr;
+
+    if (!nptr || base < 0)
+        return 0;
+
+    /* Set scanning pointer to nptr */
+    p = nptr;
+
+    /* Skip any leading whitespace */
+    while (_isspace(*p))
+        p++;
+
+    /* Handle '+' and '-' */
+    if (p[0] == '+')
+    {
+        p++;
+    }
+    else if (p[0] == '-')
+    {
+        negative = 1;
+        p++;
+    }
+
+    /* If base is zero, deduce the base from the prefix. */
+    if (base == 0)
+    {
+        if (p[0] == '0' && (p[1] == 'x' || p[1] == 'X'))
```

Comment:
Sorry for the additional complications! Technically I think, JSON does not support hex or octal ints and it allows only one leading zero, if it's before the decimal point of a float. It looks like we'll encode must numbers into strings anyways (also limited to 2^53, malleability issues, etc), but I think we should tighten the parser down to reject everything that's not in the standard.

Reply:
Thank you Christoph. We are okay because we only call it with base 10 (so that code is never executed).

```
un->integer = _strtol(start, &end, 10);
```